command/oauth/cmd.go

```diff
@@ -702,6 +765,114 @@ func (o *oauth) DoManualAuthorization() (*token, error) {
+	VerificationURL         string `json:"verification_url"`
+	VerificationURIComplete string `json:"verification_uri_complete"`
+	ExpiresIn               int    `json:"expires_in"`
+	Interval                int    `json:"interval"`
+}
+
+// DoDeviceAuthorization gets a token from the IDP using the OAuth 2.0
+// Device Authorization Grant. https://datatracker.ietf.org/doc/html/rfc8628
+func (o *oauth) DoDeviceAuthorization() (*token, error) {
+	// Identify the Device
+	data := url.Values{}
+	data.Set("client_id", o.clientID)
+	data.Set("client_secret", o.clientSecret)
+	data.Set("scope", o.scope)
+
+	resp, err := http.PostForm(o.deviceAuthzEndpoint, data)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+	defer resp.Body.Close()
+	b, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	if resp.StatusCode >= 400 {
+		var e struct {
+			Error string
+		}
+		if err := json.NewDecoder(bytes.NewReader(b)).Decode(&e); err != nil {
+			return nil, errors.Wrapf(err, "could not parse http body: %s", string(b))
+		}
+	}
+
+	var idr identifyDeviceResponse
+	if err := json.NewDecoder(bytes.NewReader(b)).Decode(&idr); err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	switch {
+	case idr.VerificationURI != "":
+		break
+	case idr.VerificationURL != "":
+		// NOTE this is a hack for Google, because their API returns the attribute
+		// 'verification_url` rather than `verification_uri`.
+		idr.VerificationURI = idr.VerificationURL
+	default:
+		return nil, errors.Errorf("device code response from server missing 'verification_uri' parameter. http body response: %s", string(b))
+	}
+
+	fmt.Fprintln(os.Stderr, "Go to the following website:")
+	fmt.Fprintln(os.Stderr)
+	fmt.Fprintln(os.Stderr, idr.VerificationURI)
+	fmt.Fprintln(os.Stderr)
+	fmt.Fprintln(os.Stderr, "and enter the activation code below:")
+	fmt.Fprintln(os.Stderr)
+	fmt.Fprintln(os.Stderr, idr.UserCode)
+	fmt.Fprintln(os.Stderr)
+
+	// Poll the Token endpoint until the user completes the flow.
+	data = url.Values{}
+	data.Set("client_id", o.clientID)
+	data.Set("client_secret", o.clientSecret)
+	data.Set("grant_type", "urn:ietf:params:oauth:grant-type:device_code")
+	data.Set("device_code", idr.DeviceCode)
+
+	var tok token
+	for {
+		resp, err := http.PostForm(o.tokenEndpoint, data)
+		if err != nil {
+			return nil, errors.WithStack(err)
+		}
+		defer resp.Body.Close()
+		b, err := io.ReadAll(resp.Body)
+		if err != nil {
+			return nil, errors.WithStack(err)
+		}
+
+		isTokenReceived := false
+		tok = token{}
+
+		switch {
+		case resp.StatusCode == http.StatusOK:
+			if err := json.NewDecoder(bytes.NewReader(b)).Decode(&tok); err != nil {
+				return nil, errors.WithStack(err)
+			}
+			isTokenReceived = true
+		case resp.StatusCode >= http.StatusBadRequest && resp.StatusCode < http.StatusInternalServerError:
+			time.Sleep(time.Duration(idr.Interval) * time.Second)
+		default:
+			return nil, errors.New(string(b))
+		}
```

Comment:
`isTokenReceive` is not required as you can just return the token. And I think we should have a timeout.